### ISSUE TITLE
flycheck’s tooltips can be displayed sdrawkcab when the source contains bidi directional formatting characters

### ISSUE THREAD

**Issue by db48x** [flycheck/flycheck#1919]
## Checklist
- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description
Source code containing bidi directional formatting characters can mess up the direction of the error message itself.

I imagine that localized error messages might cause the same problem as well, though I haven’t any easy way to check.

## Steps to reproduce

1. Start with a source file containing tricky comments:

```rust
fn main() {
    let is_admin = false;
    /*‮ } ⁦if is_admin⁩ ⁦ begin admins only */
        println!("You are an admin.");
    /* end admins only ‮ { ⁦*/
}
```
2. Rust compiler reports errors
3. Hover over the first comment, or move point to it, to see the sdrawkcab messages. See the screenshot as well.

- [x] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior
The text direction of the error message is not affected by the content of the source code it arose from.

## Screenshots
![Screenshot from 2021-11-07 23-50-47](https://user-images.githubusercontent.com/228849/140708736-0e376688-50ad-473c-b84e-cc92c2b65fd9.png)

## The Fix
I will make a pull request for this, but for the moment here is the change I made to fix this:

```lisp
(defun flycheck-error-format-message-and-id (err &optional include-snippet)
  "Format the message and id of ERR as human-readable string.

If INCLUDE-SNIPPET is non-nil, prepend the message with a snippet
of the text that the error applies to (such text can only be
determined if the error contains a full span, not just a
beginning position)."
  (let* ((id (flycheck-error-id err))
         (fname (flycheck-error-filename err))
         (other-file-p (and fname (not (equal fname (buffer-file-name))))))
    (concat (and other-file-p (format "In %S:\n" (file-relative-name fname)))
            (and include-snippet
                 (-when-let* ((snippet (flycheck-error-format-snippet err)))
                   (flycheck--format-message "`\N{LEFT-TO-RIGHT MARK}\N{FIRST STRONG ISOLATE}%s\N{POP DIRECTIONAL ISOLATE}\N{LEFT-TO-RIGHT MARK}': " snippet)))
            (or (flycheck-error-message err)
                (format "Unknown %S" (flycheck-error-level err)))
            (and id (format " [%s]" id)))))
```

Likely the other fields ought to have the same treatment.

## System configuration
I’ve got some `lsp-mode` stuff going on as well, which isn’t reflected here.

```
Syntax checkers for buffer flycheck.el in emacs-lisp-mode:

First checker to run:

  emacs-lisp
    - may enable:    yes
    - may run:       t
    - executable:    Found at /home/db48x/bin/emacs
    - next checkers: emacs-lisp-checkdoc

Checkers that may run as part of the first checker's chain:

  emacs-lisp-checkdoc
    - may enable: yes
    - executable: Found at /home/db48x/bin/emacs

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot (package: 20210825.1804)
Emacs version:    29.0.50
System:           x86_64-pc-linux-gnu
Window system:    x
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes
https://en.wikipedia.org/wiki/Bidirectional_text
https://www.w3.org/International/questions/qa-bidi-unicode-controls

**Comment by cpitclaudel** [comment]
That's a funny one, thanks!  
Question: won't this fix break messages in right-to-left languages?
(IOW, isn't the issue that rust is giving us a broken error message?)

**Comment by db48x** [comment]
No, it doesn’t break RTL messages:

![Screenshot from 2021-11-08 01-02-40](https://user-images.githubusercontent.com/228849/140713546-aab25d9a-c69e-4e05-a635-ba0b30d90ac6.png)

The message from Rust is conveyed to Emacs via the LSP protocol, and I don’t think that it is supposed to contain any extra formatting. The client is supposed to do that as needed.

**Comment by db48x** [comment]
I want to make a pull request for this, but `make check` is reporting that the file is misformatted. It doesn’t say why, or what line number is misformatted, but looking at the source I see that it wants all lines to be 80 characters or less.

Do you have any recommendations for that? The format string is 103 characters, and using hex codes instead of character names would be less readable. And using a lot of string concatenation would allocate a lot of memory.

**Comment by cpitclaudel** [comment]
> Do you have any recommendations for that?

Making a defconst for the format string is probably the simplest.

> The message from Rust is conveyed to Emacs via the LSP protocol, and I don’t think that it is supposed to contain any extra formatting.

Does that mean that every IDE needs to wrap LSP messages in those formatting codes?

**Comment by db48x** [comment]
> > The message from Rust is conveyed to Emacs via the LSP protocol, and I don’t think that it is supposed to contain any extra formatting.
> 
> Does that mean that every IDE needs to wrap LSP messages in those formatting codes?

It depends on how the client displays the message. Flycheck is putting four pieces of information into a single string and then displaying it, so it needs to be careful.

However, if we put the information into four columns of an HTML table, then each cell would be shaped and printed independently; the bidi state from one could not affect the others and thus no special care would need to be taken. Even without a table, Emacs could arrange to shape the separately, calling into libharfbuzz (or the equivalent API on other OSes) several times instead of once. But then the tooltip would have to be defined by a list of lists of strings, where the strings in the inner list are shaped independently but printed on a single line.

> > Do you have any recommendations for that?
> 
> Making a defconst for the format string is probably the simplest.

I suppose concatenating once when the constant is defined wouldn’t hurt.

**Comment by bbatsov** [comment]
This has been fixed by wrapping error snippets in Unicode directional isolate characters (U+2068 / U+2069). Bidi formatting characters in source code no longer mess up the error tooltips. Closing.

**Comment by db48x** [comment]
Thank you!
